Sell/Product/Tests/ProductRepositoryTests.swift

```diff
@@ -0,0 +1,167 @@
+        let graphQLManager = MockGraphQLManager<DataType>()
+        let sut = ProductRepository(graphQLManager: graphQLManager)
+        return (graphQLManager: graphQLManager, sut: sut)
+    }
+
+    private func makeProductFragment(id: String = "1") -> ApolloType.FragmentProduct {
+        ApolloType.FragmentProduct(id: id)
+    }
+}
+
+extension ProductRepositoryTests {
+
+    func test_getProducts_graphQLManagerHasData_completeWithSuccessCase() {
+        let (graphQLManager, sut) = makeGraphQLManagerAndSUT(ofType: ApolloType.GetProductsQuery.Data.self)
+
+        let getProducts = ApolloType
+            .GetProductsQuery
+            .Data
+            .GetProduct(edges: [], nextPage: 123)
+        let data = ApolloType
+            .GetProductsQuery
+            .Data(getProducts: getProducts)
+        let expectedResult = GraphQLResult<
+            ApolloType
+            .GetProductsQuery
+            .Data>(
+            data: data,
+            extensions: nil,
+            errors: nil,
+            source: .cache,
+            dependentKeys: nil)
+        graphQLManager.resultReturns = .success(expectedResult)
+
+        let expectation = XCTestExpectation(description: "callback is fulfilled")
+        sut.getProducts(
+            pageSize: 1,
+            page: 1,
+            filters: nil
+        ) { result in
+            guard case let .success(returnedResponse) = result else {
+                return XCTFail("This test should return success")
+            }
+            XCTAssertEqual(
+                returnedResponse.items.count,
+                expectedResult.data?.getProducts?.edges?.count)
+            XCTAssertEqual(
+                returnedResponse.nextPage,
+                expectedResult.data?.getProducts?.nextPage)
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 0.01)
+    }
+
+    func test_getProducts_graphQLManagerReturnsError_completeWithFailureCase() {
+        let (graphQLManager, sut) = makeGraphQLManagerAndSUT(ofType: ApolloType.GetProductsQuery.Data.self)
+
+        graphQLManager.resultReturns = .failure(DummyError.failure)
+        let expectation = XCTestExpectation(description: "callback is fulfilled")
+        sut.getProducts(
+            pageSize: 1,
+            page: 1,
+            filters: nil
+        ) { result in
+            guard case let .failure(returnedError) = result else {
+                return XCTFail("This test should return failure")
+            }
+            XCTAssertEqual(returnedError as? DummyError, DummyError.failure)
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 0.01)
+    }
+
+    func test_getProductById_graphQLManagerHasData_completeWithSuccessCase() {
+        let (graphQLManager, sut) = makeGraphQLManagerAndSUT(ofType: ApolloType.GetProductByIdQuery.Data.self)
+        let productFragment = makeProductFragment()
```

Comment:
Just curios why choose to call another function here, instead using `ApolloType.FragmentProduct(id: "1")` directly please?

Reply:
I thought I would use the helper for other cases, but obviously, I didn't. I will change it to a constant.